File: lua/lc-lookup.lua
-- LoanConnect lookup-table picker (lenders, partners, products, ...).
--
-- Backend: bin/lc-lookup.php — dumps tables defined in
-- ~/.config/lc-lookup/manifest.json to JSON files in ~/.cache/lc-lookup/.
-- This module reads those JSON files and feeds snacks.picker.
--
-- Usage:
--   :LcLookup                 -- pick a table, then a row
--   :LcLookup lenders         -- jump straight to lenders
--   :LcLookupRefresh          -- async dump-all
--   :LcLookupRefresh lenders  -- async dump one
--
-- On row select: yanks "<id>" into register `c` and the system clipboard,
-- and notifies the chosen row. Same convention as lc-cyber.lua.

local M = {}

local script_path = vim.fn.stdpath 'config' .. '/bin/lc-lookup.php'

local function cache_dir()
  local override = vim.env.LC_LOOKUP_CACHE
  if override and override ~= '' then
    return override
  end
  local xdg = vim.env.XDG_CACHE_HOME
  if xdg and xdg ~= '' then
    return xdg .. '/lc-lookup'
  end
  return vim.fn.expand '~/.cache/lc-lookup'
end

---Read a cached lookup file and return its decoded payload, or nil + error.
---@param table_name string
---@return table|nil payload
---@return string|nil err
local function read_cache(table_name)
  local path = cache_dir() .. '/' .. table_name .. '.json'
  local fd = io.open(path, 'r')
  if not fd then
    return nil, 'no cache for "' .. table_name .. '" — run :LcLookupRefresh ' .. table_name
  end
  local raw = fd:read '*a'
  fd:close()
  local ok, decoded = pcall(vim.json.decode, raw)
  if not ok or type(decoded) ~= 'table' or type(decoded.rows) ~= 'table' then
    return nil, 'cache for "' .. table_name .. '" is malformed (' .. path .. ')'
  end
  return decoded, nil
end

---List available cached tables (filenames in cache dir).
---@return string[]
local function list_cached_tables()
  local dir = cache_dir()
  local out = {}
  for name, _ in vim.fs.dir(dir) do
    local stem = name:match '^(.+)%.json$'
    if stem then
      table.insert(out, stem)
    end
  end
  table.sort(out)
  return out
end

---Yank a value to register `c` and the system clipboard, with a notification.
---@param value string
---@param label string
local function yank_with_notify(value, label)
  vim.fn.setreg('c', value)
  vim.fn.setreg('+', value)
  vim.notify(string.format('%s = %s  (yanked to "c and clipboard)', label, value), vim.log.levels.INFO)
end

---Pick from a table's rows. Each row should have at least an id-ish and label-ish field;
---we autodetect them. Pass `field_id` / `field_label` in opts to override.
---@param table_name string
---@param opts? { field_id?: string, field_label?: string }
function M.pick(table_name, opts)
  opts = opts or {}
  local payload, err = read_cache(table_name)
  if not payload then
    vim.notify(err, vim.log.levels.ERROR)
    return
  end

  local rows = payload.rows
  if #rows == 0 then
    vim.notify('lc-lookup: "' .. table_name .. '" cache is empty', vim.log.levels.WARN)
    return
  end

  -- Autodetect id / label columns from the first row's keys, unless overridden.
  local sample = rows[1]
  local field_id = opts.field_id
  local field_label = opts.field_label
  if not field_id then
    for _, k in ipairs { 'id', 'ID', 'uuid' } do
      if sample[k] ~= nil then
        field_id = k
        break
      end
    end
  end
  if not field_label then
    for _, k in ipairs { 'name', 'label', 'title', 'display_name', 'company_name' } do
      if sample[k] ~= nil then
        field_label = k
        break
      end
    end
  end
  if not field_id or not field_label then
    vim.notify(
      ('lc-lookup: cannot autodetect id/label columns in "%s" (have: %s). Pass field_id/field_label.'):format(
        table_name,
        table.concat(vim.tbl_keys(sample), ', ')
      ),
      vim.log.levels.ERROR
    )
    return
  end

  local items = {}
  for _, row in ipairs(rows) do
    local id_str = tostring(row[field_id])
    local label_str = tostring(row[field_label] or '')
    table.insert(items, {
      id = id_str,
      label = label_str,
      text = id_str .. '  ' .. label_str, -- snacks.picker fuzzy-matches on `text`
      row = row,
    })
  end

  local refreshed = payload.refreshed_at and (' (refreshed ' .. payload.refreshed_at .. ')') or ''
  local title = ('lc-lookup: %s — %d rows%s'):format(table_name, #rows, refreshed)

  local ok, snacks = pcall(require, 'snacks')
  if not ok or not snacks.picker then
    -- Fallback: vim.ui.select. Always works, no plugin dependency.
    local choices = vim.tbl_map(function(it)
      return it.text
    end, items)
    vim.ui.select(choices, { prompt = title }, function(_, idx)
      if idx then
        yank_with_notify(items[idx].id, table_name .. '.' .. items[idx].label)
      end
    end)
    return
  end

  snacks.picker.pick {
    title = title,
    items = items,
    format = function(item)
      return {
        { ('%-8s '):format(item.id), 'Number' },
        { item.label, 'Identifier' },
      }
    end,
    confirm = function(picker, item)
      picker:close()
      if item then
        yank_with_notify(item.id, table_name .. '.' .. item.label)
      end
    end,
  }
end

---Pick a table first, then a row.
function M.pick_any()
  local tables = list_cached_tables()
  if #tables == 0 then
    vim.notify('lc-lookup: no cached tables in ' .. cache_dir() .. '. Run :LcLookupRefresh', vim.log.levels.WARN)
    return
  end
  if #tables == 1 then
    return M.pick(tables[1])
  end
  vim.ui.select(tables, { prompt = 'lc-lookup: which table?' }, function(choice)
    if choice then
      M.pick(choice)
    end
  end)
end

---Async refresh via `bin/lc-lookup.php`. Notifies on completion.
---@param table_name? string  nil => dump-all
function M.refresh(table_name)
  local cmd = { script_path }
  if table_name and table_name ~= '' then
    table.insert(cmd, 'dump')
    table.insert(cmd, table_name)
  else
    table.insert(cmd, 'dump-all')
  end

  vim.notify('lc-lookup: refreshing ' .. (table_name or 'all tables') .. ' ...', vim.log.levels.INFO)

  vim.system(cmd, { text = true }, function(out)
    vim.schedule(function()
      if out.code == 0 then
        vim.notify('lc-lookup: refresh OK\n' .. (out.stderr or ''):gsub('%s+$', ''), vim.log.levels.INFO)
      else
        vim.notify('lc-lookup: refresh FAILED (exit ' .. out.code .. ')\n' .. (out.stderr or out.stdout or ''), vim.log.levels.ERROR)
      end
    end)
  end)
end

---Smoke-test the DB connection synchronously.
function M.check()
  local result = vim.system({ script_path, 'check' }, { text = true }):wait()
  if result.code == 0 then
    vim.notify('lc-lookup check: ' .. (result.stderr or ''):gsub('%s+$', ''), vim.log.levels.INFO)
  else
    vim.notify('lc-lookup check FAILED:\n' .. (result.stderr or result.stdout or ''), vim.log.levels.ERROR)
  end
end

return M

File: bin/lc-lookup.php
#!/usr/bin/env php
<?php

/**
 * Standalone lookup-table dumper for LoanConnect (and similar) MySQL data.
 *
 * Companion to bin/lc-codec.php. Reads a manifest of "table-name => SELECT SQL"
 * entries and writes each result set as JSON into a cache directory. The
 * Neovim side (lua/lc-lookup.lua) reads those JSON files and feeds a picker.
 *
 * Why standalone (no Laravel): same reason as lc-codec.php — the lookup
 * picker has to work from any machine that can reach the DB, without
 * cloning the loanconnect repo. If you need Eloquent accessors / model
 * mutators, make a parallel artisan command in the LoanConnect project.
 *
 * Usage:
 *   lc-lookup.php list                            # list tables in manifest
 *   lc-lookup.php dump <table> [--out=<dir>]      # dump one table
 *   lc-lookup.php dump-all [--out=<dir>]          # dump all tables
 *   lc-lookup.php check                           # smoke-test the connection
 *
 * Connection (first match wins):
 *   1. $LC_DB_DSN  (full PDO DSN, e.g. "mysql:host=127.0.0.1;port=3306;dbname=loanconnect")
 *      with $LC_DB_USER / $LC_DB_PASS for credentials
 *   2. $LC_DB_HOST / $LC_DB_PORT / $LC_DB_NAME / $LC_DB_USER / $LC_DB_PASS
 *   3. ~/.config/lc-lookup/connection.json — { "dsn": "...", "user": "...", "pass": "..." }
 *      (chmod 600 it; this script refuses to read world-readable connection files)
 *
 * Manifest:
 *   ~/.config/lc-lookup/manifest.json — see docs/lc-lookup.md for format.
 *   Override with $LC_LOOKUP_MANIFEST.
 *
 * Cache (default --out):
 *   $XDG_CACHE_HOME/lc-lookup/ (falls back to ~/.cache/lc-lookup/)
 *   Override with $LC_LOOKUP_CACHE.
 *
 * Output contract (programmatic callers):
 *   stdout: nothing in dump modes (writes files); JSON listing in `list` mode
 *   stderr: errors and progress
 *   exit 0 on success, 1 on connection/SQL failure, 2 on usage error
 */
const VERSION = '1.0.0';

function fatal(string $msg, int $code = 1): never
{
    fwrite(STDERR, "lc-lookup: $msg\n");
    exit($code);
}

function info(string $msg): void
{
    fwrite(STDERR, "lc-lookup: $msg\n");
}

function home_dir(): string
{
    $home = getenv('HOME') ?: posix_getpwuid(posix_getuid())['dir'] ?? null;
    if (! $home) {
        fatal('cannot determine home directory');
    }

    return $home;
}

function config_dir(): string
{
    $xdg = getenv('XDG_CONFIG_HOME');

    return $xdg ? "$xdg/lc-lookup" : home_dir().'/.config/lc-lookup';
}

function default_cache_dir(): string
{
    if ($override = getenv('LC_LOOKUP_CACHE')) {
        return $override;
    }
    $xdg = getenv('XDG_CACHE_HOME');

    return $xdg ? "$xdg/lc-lookup" : home_dir().'/.cache/lc-lookup';
}

function manifest_path(): string
{
    return getenv('LC_LOOKUP_MANIFEST') ?: config_dir().'/manifest.json';
}

function load_manifest(): array
{
    $path = manifest_path();
    if (! is_readable($path)) {
        fatal("manifest not found: $path (see docs/lc-lookup.md)", 2);
    }
    $raw = file_get_contents($path);
    $data = json_decode($raw, true);
    if (! is_array($data) || ! isset($data['tables']) || ! is_array($data['tables'])) {
        fatal("manifest malformed: expected top-level { \"tables\": { ... } } in $path", 2);
    }
    foreach ($data['tables'] as $name => $entry) {
        if (! is_array($entry) || empty($entry['sql'])) {
            fatal("manifest entry '$name' missing 'sql' field", 2);
        }
    }

    return $data['tables'];
}

function load_connection(): array
{
    // Mode 1: explicit DSN env
    if ($dsn = getenv('LC_DB_DSN')) {
        return [
            'dsn' => $dsn,
            'user' => getenv('LC_DB_USER') ?: null,
            'pass' => getenv('LC_DB_PASS') ?: null,
        ];
    }

    // Mode 2: per-component env
    $host = getenv('LC_DB_HOST');
    $name = getenv('LC_DB_NAME');
    if ($host && $name) {
        $port = getenv('LC_DB_PORT') ?: '3306';

        return [
            'dsn' => "mysql:host=$host;port=$port;dbname=$name;charset=utf8mb4",
            'user' => getenv('LC_DB_USER') ?: null,
            'pass' => getenv('LC_DB_PASS') ?: null,
        ];
    }

    // Mode 3: connection.json file
    $path = config_dir().'/connection.json';
    if (is_readable($path)) {
        // Refuse to read world-readable files — this holds passwords.
        $perms = fileperms($path) & 0o777;
        if ($perms & 0o077) {
            fatal(sprintf('connection.json is mode %04o; chmod 600 it', $perms), 2);
        }
        $cfg = json_decode(file_get_contents($path), true);
        if (! is_array($cfg) || empty($cfg['dsn'])) {
            fatal('connection.json malformed: needs at least { "dsn": "..." }', 2);
        }

        return [
            'dsn' => $cfg['dsn'],
            'user' => $cfg['user'] ?? null,
            'pass' => $cfg['pass'] ?? null,
        ];
    }

    fatal('no connection configured. Set LC_DB_DSN or LC_DB_HOST/LC_DB_NAME, or create '.config_dir().'/connection.json (chmod 600)', 2);
}

function connect(): PDO
{
    $cfg = load_connection();
    try {
        $pdo = new PDO($cfg['dsn'], $cfg['user'], $cfg['pass'], [
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
            PDO::ATTR_EMULATE_PREPARES => false,
            PDO::ATTR_TIMEOUT => 5,
        ]);

        return $pdo;
    } catch (PDOException $e) {
        fatal('connection failed: '.$e->getMessage());
    }
}

function ensure_dir(string $dir): void
{
    if (is_dir($dir)) {
        return;
    }
    if (! @mkdir($dir, 0o755, true) && ! is_dir($dir)) {
        fatal("cannot create cache dir: $dir");
    }
}

function dump_table(PDO $pdo, string $name, array $entry, string $outDir): int
{
    $sql = $entry['sql'];
    info("dumping '$name' ...");
    $start = microtime(true);
    try {
        $rows = $pdo->query($sql)->fetchAll();
    } catch (PDOException $e) {
        fatal("query failed for '$name': ".$e->getMessage());
    }
    $payload = [
        'table' => $name,
        'refreshed_at' => date('c'),
        'source_sql' => $sql,
        'row_count' => count($rows),
        'rows' => $rows,
    ];
    $path = "$outDir/$name.json";
    $tmp = "$path.tmp";
    if (file_put_contents($tmp, json_encode($payload, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT)) === false) {
        fatal("cannot write $tmp");
    }
    // Atomic move so a partial write never replaces a good cache file.
    if (! rename($tmp, $path)) {
        fatal("cannot rename $tmp -> $path");
    }
    $elapsed = sprintf('%.2fs', microtime(true) - $start);
    info(sprintf('  -> %d rows in %s (%s)', count($rows), $elapsed, $path));

    return count($rows);
}

function parse_opts(array $argv): array
{
    $opts = ['out' => null];
    $rest = [];
    for ($i = 1; $i < count($argv); $i++) {
        $a = $argv[$i];
        if (str_starts_with($a, '--out=')) {
            $opts['out'] = substr($a, 6);
        } else {
            $rest[] = $a;
        }
    }

    return [$opts, $rest];
}

function usage(int $code = 2): never
{
    $version = VERSION;
    fwrite(STDERR, <<<TXT
        lc-lookup.php v$version — dump MySQL lookup tables to JSON for snacks.picker

        usage:
          lc-lookup.php list
          lc-lookup.php dump <table> [--out=<dir>]
          lc-lookup.php dump-all [--out=<dir>]
          lc-lookup.php check

        env:
          LC_DB_DSN / LC_DB_HOST / LC_DB_NAME / LC_DB_USER / LC_DB_PASS
          LC_LOOKUP_MANIFEST  (default: ~/.config/lc-lookup/manifest.json)
          LC_LOOKUP_CACHE     (default: ~/.cache/lc-lookup)


        TXT);
    exit($code);
}

[$opts, $args] = parse_opts($argv);
$cmd = $args[0] ?? null;

switch ($cmd) {
    case 'list':
        $tables = load_manifest();
        echo json_encode([
            'manifest' => manifest_path(),
            'tables' => array_keys($tables),
        ], JSON_PRETTY_PRINT)."\n";
        exit(0);

    case 'check':
        $pdo = connect();
        $stmt = $pdo->query('SELECT VERSION() AS v, DATABASE() AS db, NOW() AS now');
        $row = $stmt->fetch();
        info(sprintf('OK: MySQL %s, db=%s, server-time=%s', $row['v'], $row['db'] ?? 'NULL', $row['now']));
        exit(0);

    case 'dump':
        if (empty($args[1])) {
            usage();
        }
        $tables = load_manifest();
        $name = $args[1];
        if (! isset($tables[$name])) {
            fatal("unknown table '$name'. Available: ".implode(', ', array_keys($tables)), 2);
        }
        $outDir = $opts['out'] ?: default_cache_dir();
        ensure_dir($outDir);
        $pdo = connect();
        dump_table($pdo, $name, $tables[$name], $outDir);
        exit(0);

    case 'dump-all':
        $tables = load_manifest();
        $outDir = $opts['out'] ?: default_cache_dir();
        ensure_dir($outDir);
        $pdo = connect();
        $total = 0;
        foreach ($tables as $name => $entry) {
            $total += dump_table($pdo, $name, $entry, $outDir);
        }
        info("done. $total rows total across ".count($tables).' tables.');
        exit(0);

    default:
        usage();
}
